[RLlib] Cont: Contribution of LeelaChessZero algorithm for playing chess in a MultiAgent env

Capiru opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
@sven1977 
This is a continuation of : https://github.com/ray-project/ray/pull/30651 due to unexpected behaviour in the PR

This implements a version of LeelaChessZero, largely inspired by AlphaZero but with some structural changes. This changes also enable the slightly changed MCTS in the LeelaChessZero folder to run on MultiAgent environments. LeelaChessZero comes with an implementation of a MultiAgentChess env and also a test for making sure it builds and runs on the appropriate chess environment.

## Related issue number

no related issue

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests: added a test to the new algorithm folder that runs successfully locally, might require "pip install chess"
   - [x] Release tests
   - [ ] This PR is not tested :(

Capiru commented:
After fixing the latest upgrade of the pettingzoo upgrade to a newer version, the only failing test related to the contribution is the failed link check that should work after the merge of this PR.

kouroshHakha commented on rllib/algorithms/leela_chess_zero/__init__.py:
```diff
@@ -0,0 +1,17 @@
+    LeelaChessZero,
+    LeelaChessZeroConfig,
+    LeelaChessZeroDefaultCallbacks,
+    DEFAULT_CONFIG,
```
no need to import DEFAULT_CONFIG again. We have moved away from dict configs and are using AlgorithmConfigs now.

kouroshHakha commented on rllib/algorithms/leela_chess_zero/leela_chess_zero.py:
```diff
@@ -0,0 +1,414 @@
+        return train_results
+
+
+DEFAULT_CONFIG = LeelaChessZeroConfig().to_dict()
```
This is deprecated. Thanks

kouroshHakha commented on rllib/algorithms/leela_chess_zero/mcts.py:
```diff
@@ -0,0 +1,214 @@
+"""
+Mcts implementation modified from
```
Very cool

kouroshHakha commented on rllib/examples/env/pettingzoo_chess.py:
```diff
@@ -0,0 +1,193 @@
+    def __init__(
+        self,
+        config: Dict[Any, Any] = None,
+        env: AECEnv = chess_v5(),
```
Can you make sure your constructor does not need the env object to be passed in? This will create issue for serializing the environment. One thing you could do is have config include a base env key in it and if None, default to chess_v5() inside the constructor. Thanks.

kouroshHakha commented on rllib/algorithms/leela_chess_zero/leela_chess_zero_model.py:
```diff
@@ -0,0 +1,174 @@
+import torch.nn as nn
```
use try_import_torch()

kouroshHakha reviewed (commented):
Hey @Capiru, Thanks for such a great PR. I left a couple of minor comments. Also one of the unittests which checks for the failure of rllib import when torch is not installed is failing. You need to make sure you use try_torch_import API whereever you need `import torch`. Ping me when all the tests pass and you have addressed the concerns. I'll get a commiter to merge your PR. cc @sven1977 @gjoliver

Capiru commented:
> Hey @Capiru, Thanks for such a great PR. I left a couple of minor comments. Also one of the unittests which checks for the failure of rllib import when torch is not installed is failing. You need to make sure you use try_torch_import API whereever you need `import torch`. Ping me when all the tests pass and you have addressed the concerns. I'll get a commiter to merge your PR. cc @sven1977 @gjoliver

Thanks for the comments, I believe now all relevant tests are passing, and all comments were addressed. @kouroshHakha

sven1977 reviewed (approved):
Thanks a lot for this awesome PR @Capiru !
Looks great and ready to merge.
